File: typescript/src/prototype/prototype.ts
interface Cloneable<T> {
  clone(): T;
}

// I am nesting for the sake of nesting
type Data = {
  dataMap: Map<string, any>
}

// Dataframe takes data in the constructor which is an object type so passed by reference
export class DataFrame implements Cloneable<DataFrame> {
  name: string
  data: Data

  constructor(name: string, data: Data) {
    this.name = name
    this.data = data
  }

  clone(): DataFrame {
    // structuredClone is a native function that clones objects
    // If you want to use something else, go ahead like lodash
    // or your own implementation, which will be easy for this example
    return structuredClone(this)
  }
}

File: go/prototype/prototype.go
package prototype

// Cloneable interface or should it be Cloner or something like that? Go idioms help!
type Cloneable[T any] interface {
	Clone() T
}

// I don't know how data people do things, so the struct might be a bit off.
//But it is just an example to understand Prototype Design pattern...

// Data for our example
type Data struct {
	Map map[string]any
}

func NewData(m map[string]any) *Data {
	return &Data{
		Map: m,
	}
}

// DataFrame takes pointer to the Data struct, to make things a bit harder by nesting stuff...
type DataFrame struct {
	Name string
	Data *Data
}

// NewDataFrame constructs new DataFrame
func NewDataFrame(name string, data *Data) *DataFrame {
	return &DataFrame{
		Name: name,
		Data: data,
	}
}

// Clone method for DataFrame, to implement Cloneable interface
func (d DataFrame) Clone() DataFrame {
	// We need to clone map, not just copy the reference of it!
	deepCloneMap := make(map[string]any)
	for k, v := range d.Data.Map {
		deepCloneMap[k] = v
	}

	// Primitives are just copied, no need to clone them
	return DataFrame{
		Name: d.Name,
		Data: &Data{
			Map: deepCloneMap,
		},
	}
}
